common/lib/xmodule/xmodule/partitions/partitions_service.py

```diff
@@ -3,41 +3,114 @@
 user partitions.  It uses the user_service key/value store provided by the LMS runtime to
 persist the assignments.
 """
-from abc import ABCMeta, abstractproperty
+from django.conf import settings
+from django.utils.translation import ugettext_lazy as _
+import logging
+
+from xmodule.partitions.partitions import UserPartition, UserPartitionError
+from xmodule.modulestore.django import modulestore
+
+
+log = logging.getLogger(__name__)
+
+
+# UserPartition IDs must be unique. The Cohort and Random UserPartitions (when they are
+# created via Studio) choose an unused ID in the range of 100 to MAX_INT. Therefore the
+# dynamic UserPartitionIDs must be under 100, and they have to be hard-coded to ensure
+# they are always the same whenever the dynamic partition is added (since the UserPartition
+# ID is stored in the xblock group_access dict).
+ENROLLMENT_TRACK_PARTITION_ID = 50
+
+
+# settings will not be available when running nosetests.
+FEATURES = getattr(settings, 'FEATURES', {})
+
+
+def get_all_partitions_for_course(course):
+    """
+    A method that returns all `UserPartitions` associated with a course, as a List.
+    This will include the ones defined in course.user_partitions, but it may also
+    include dynamically included partitions (such as the `EnrollmentTrackUserPartition`).
+    """
+    return course.user_partitions + _get_dynamic_partitions(course)
+
+
+def _get_dynamic_partitions(course):
+    """
+    Return the dynamic user partitions for this course.
+    If none exists, returns an empty array.
+    """
+    enrollment_partition = _create_enrollment_track_partition(course)
+    return [enrollment_partition] if enrollment_partition else []
+
+
+def _create_enrollment_track_partition(course):
+    """
+    Create and return the dynamic enrollment track user partition.
+    If it cannot be created, None is returned.
+    """
+    if not FEATURES.get('ENABLE_ENROLLMENT_TRACK_USER_PARTITION'):
+        return None
+
+    try:
+        enrollment_track_scheme = UserPartition.get_scheme("enrollment_track")
+    except UserPartitionError:
+        log.warning("No 'enrollment_track' scheme registered, EnrollmentTrackUserPartition will not be created.")
+        return None
+
+    used_ids = set(p.id for p in course.user_partitions)
```

Comment:
@nasthagiri I have thought more about this and am coming around to an exception. We are guaranteed that partitions created in Studio won't conflict (they are > 100), but partitions created in some other way could potentially conflict.

I like option #2. I will create a ticket for it in our epic.

Note that when we get to the discussions work we will likely want to throw an exception if 2 user partitions exist with a group that has the same ID. That's another place where we should be fine for partitions created by Studio, but could in the future run into difficulty.